### ISSUE TITLE
 Spurious rebuilds when VSCode opened through symlink

### ISSUE THREAD

**Issue by cstrahan-blueshift** [rust-lang/rust-analyzer#13593]
If VSCode is opened via something like:

```
$ code some_symlink_to_project
```

saving any Rust source file in the project (presumably it's the `cargo check` functionality being triggered) will result in spurious rebuilds when you run `cargo build` from the command line, despite the dependencies being unchanged.

I have a full, turn-key reproducer here: https://github.com/cstrahan-blueshift/vscode_rust_rebuild_repro

---

**rust-analyzer version**: 0.3.1266-standalone

**rustc version**: rustc 1.65.0 (897e37553 2022-11-02)

**relevant settings**: N/A

**Comment by lnicola** [comment]
That's a great analysis you've done there and it shows what triggers the rebuilds: https://github.com/nbigaouette/onnxruntime-rs/blob/master/onnxruntime-sys/build.rs#L389. We're passing a non-canonical path to the symlink in https://github.com/rust-lang/rust-analyzer/blob/d03c1c87d4ca2d524646316387d47b12524ac451/crates/project-model/src/build_scripts.rs#L322-L329. Perhaps `AbsPathBuf` should only contain canonical paths? I don't know.

But not even `cargo` handles this perfectly:

```
$ cd ~/topdir/vscode_rust_rebuild_repro
$ cargo clean
$ cargo check
   Compiling libc v0.2.137
   Compiling cc v1.0.76
   Compiling proc-macro2 v1.0.47
   Compiling unicode-ident v1.0.5
   Compiling quote v1.0.21
   Compiling cfg-if v1.0.0
   Compiling syn v1.0.103
   Compiling once_cell v1.16.0
   Compiling untrusted v0.7.1
   Compiling spin v0.5.2
   Compiling autocfg v1.1.0
   Compiling pkg-config v0.3.26
   Compiling crc32fast v1.3.2
   Compiling tinyvec_macros v0.1.0
   Compiling log v0.4.17
   Compiling adler v1.0.2
   Compiling thiserror v1.0.37
   Compiling percent-encoding v2.2.0
   Compiling unicode-bidi v0.3.8
   Compiling rustls v0.20.7
   Compiling base64 v0.13.1
   Compiling byteorder v1.4.3
   Compiling chunked_transfer v1.4.0
    Checking rawpointer v0.2.1
    Checking pin-project-lite v0.2.9
    Checking lazy_static v1.4.0
   Compiling miniz_oxide v0.5.4
   Compiling num-traits v0.2.15
   Compiling num-integer v0.1.45
    Checking tracing-core v0.1.30
    Checking matrixmultiply v0.3.2
   Compiling tinyvec v1.6.0
   Compiling form_urlencoded v1.1.0
   Compiling ring v0.16.20
   Compiling bzip2-sys v0.1.11+1.0.8
   Compiling unicode-normalization v0.1.22
   Compiling flate2 v1.0.24
   Compiling idna v0.3.0
   Compiling xattr v0.2.3
   Compiling filetime v0.2.18
   Compiling time v0.1.44
   Compiling tar v0.4.38
   Compiling bzip2 v0.4.3
   Compiling url v2.3.1
   Compiling thiserror-impl v1.0.37
   Compiling tracing-attributes v0.1.23
    Checking num-complex v0.4.2
    Checking ndarray v0.15.1
   Compiling webpki v0.22.0
   Compiling sct v0.7.0
   Compiling webpki-roots v0.22.5
   Compiling ureq v2.5.0
    Checking tracing v0.1.37
   Compiling zip v0.5.13
   Compiling onnxruntime-sys v0.0.14
    Checking onnxruntime v0.0.14
    Checking demo_lib v0.1.0 (~/topdir/vscode_rust_rebuild_repro/crates/demo_lib)
    Checking demo_bin v0.1.0 (~/topdir/vscode_rust_rebuild_repro/crates/demo_bin)
    Finished dev [unoptimized + debuginfo] target(s) in 4.34s
$ cd ~/vscode_rust_rebuild_repro_link
$ cargo check
   Compiling onnxruntime-sys v0.0.14
    Checking onnxruntime v0.0.14
    Checking demo_lib v0.1.0 (~/topdir/vscode_rust_rebuild_repro/crates/demo_lib)
    Checking demo_bin v0.1.0 (~/topdir/vscode_rust_rebuild_repro/crates/demo_bin)
    Finished dev [unoptimized + debuginfo] target(s) in 0.32s
```

I haven't tested with RA, but after going back to `~/topdir/vscode_rust_rebuild_repro`, `cargo check` won't build it a third time.

**Comment by lnicola** [comment]
Yeah, I see the same thing without involving RA:

![image](https://user-images.githubusercontent.com/308347/201049403-4d7d3697-a597-4f6f-986d-1764f3ed272d.png)

![image](https://user-images.githubusercontent.com/308347/201049467-962268a3-6e5b-4dda-8242-f9b73e47a964.png)

**Comment by lnicola** [comment]
Closing, feel free to comment here if you disagree with my conclusion that this is a cargo "bug" (as shown in the previous comment).

**Comment by cstrahan-blueshift** [comment]
Thanks @lnicola for looking at this.

I think I disagree with your conclusion: try `touch`ing the source file (while `cd`ed into the symlink in your shell) and run `cargo build` again. In full, what I find (all where `pwd` is the symlink path):

1. Run `cargo build` in my shell, get a full rebuild
2. Run `cargo build` again, and see that the cache was warm -- build is effectively a no-op
3. I save the file in vscode, triggering RA
4. RA invalidates the cache that the previous `cargo build` created
5. Run `cargo build` in my shell, get a full rebuild as a consequence of the previous point
6. Run `cargo build` again, and see that the cache was warm -- build is effectively a no-op
7. Run `touch crates/demo_bin/src/main.rs`
8. Run `cargo build` in my shell, and find that `demo_lib` does _not_ end up being rebuilt, only `demo_bin` is rebuilt.
9. Run `cargo build` again, and see that the cache was warm -- build is effectively a no-op

I _think_ the last screenshot you shared was essentially step 5 above. Could you confirm that?

**Comment by lnicola** [comment]
> I think the last screenshot you shared was essentially step 5 above. Could you confirm that?

No, it was exactly what you see in the screenshots. A `cargo build` through the symlink after a `cargo build` in the project directory. RA or Code were not involved, nor did I update any file.

I think a `cargo build` won't _always_ rebuild, though. It says something about the timestamps.

**Comment by lnicola** [comment]
> Just to check (correct me if I'm wrong), I see that you're concerned with this:

 - cargo build, finding that the project is rebuilt in full

Not in full, just `onnxruntime-sys`, `onnxruntime`, and the project crates (since their dependency was rebuilt), but yes.

> Which is in contrast with my concern:

 - Open a terminal emulator (perhaps even using vscode's built in terminal emulator, which conveniently should spawn a shell in ~/vscode_rust_rebuild_repro_link)

Only if I run it like that. `cd ~/vscode_rust_rebuild_repro_link; code .` works fine because it opens the symlink destination.

 - Run cargo build, and find that a complete rebuild is performed

Confirmed.

But I still don't think this matters too much. As I've shown before, `cargo` itself is inconsistent. You can even see it here if you look at the crate path:

![image](https://user-images.githubusercontent.com/308347/201424572-645cf965-8ebf-4c76-8a7b-f96ffb73daa9.png)

**Comment by cstrahan-blueshift** [comment]
> cargo build, finding that the project is rebuilt in full

Correct; I was imprecise. In my case (on my project at work) our crate that depends on `onnxruntime` sits at the bottom of our dependency graph, and thus results in (at the very least) very time consuming linking of all other crates. So (nearly) a full rebuild in our case, but not as general rule, no.

> Only if I run it like that. cd ~/vscode_rust_rebuild_repro_link; code . works fine because it opens the symlink destination.

Ah. I tend to lean heavily on fuzzy history search in my shell (using `fzf`, specifically), so I tend to `code ~/src/some_project` knowing that I can (from any other shell I have open) quickly do something like `Ctrl+R co smpjct`, which effectively expands to `code ~/src/some_project`, which in turn does _not_ follow the symlink to its destination.

> But I still don't think this matters too much. As I've shown before, cargo itself is inconsistent. You can even see it here if you look at the crate path:

Hmmm. Inconsistent in which way? Sorry if I'm being dense. `cargo check` and `cargo build` at least seem to be consistent in using the canonical path (where `~/src/plasma_rust` is the symlink -- this is the work project where I stumbled across this disparity between RA and `cargo`):

![image](https://user-images.githubusercontent.com/85186038/201426238-df99708c-3c02-431e-a7c5-64b77e4476a8.png)

Unless you're just referring to the inconsistency between the effective path used by check/build vs the path that is the cwd (i.e. `/home/cstrahan/src/firefly/vendor/gems/plasma_rust/crates/plasma` vs `/home/cstrahan/src/plasma_rust`).
